File: AchtungLunneSpa/Scripts/Game/Game.ts
import ko = require("knockout");
import ClientPlayer = require("./ClientPlayer");
import ClientMap = require("./ClientMap");
import KeyboardGroup = require("./KeyboardGroup");
import Rendering = require("./Rendering");
import Team = require("./Team");
import toastr = require("toastr");
import $ = require('jquery');
import utils = require('../common/Utils');
import moment = require('moment');
import NetworkHandler = require('./NetworkHandler');

class Game {

    ctx: CanvasRenderingContext2D;

	currentLocalPlayers: KnockoutObservableArray<ClientPlayer>;
    currentPlayers: KnockoutObservableArray<ClientPlayer>;
    currentMap: KnockoutObservable<ClientMap>;
    gameOn: KnockoutObservable<boolean>;
    networkHandler: NetworkHandler;

    stopMain: number;
    lastRender: number;
    lastTick: number = performance.now();
    tickLength: number = 50;

	myHub: MyHub;
	connectionID: string;

    private textArea: KnockoutObservable<string>;

    constructor(ctx: CanvasRenderingContext2D, textArea: KnockoutObservable<string>) {
		this.currentLocalPlayers = ko.observableArray<ClientPlayer>();
        this.currentPlayers = ko.observableArray<ClientPlayer>();
        this.currentMap = ko.observable<ClientMap>();
        this.gameOn = ko.observable<boolean>();
        this.lastRender = this.lastTick;
        this.textArea = textArea;

        this.ctx = ctx;
        if (this.ctx != null) {
            this.initCanvas();
			this.searchForGame();
        } else {
            alert("Cannot find canvas!");
        }
    }

	onBlur() {
		var players = this.currentPlayers();
		for (var i = 0; i < players.length; i++) {
            if (players[i].keyboardStates != null) {
                players[i].keyboardStates.resetAll();
            }
        }
	}

    onKeyDown(e: KeyboardEvent) {
        if (!this.gameOn()) {
            return;
        }

        var players = this.currentLocalPlayers();
        for (var i = 0; i < players.length; i++) {
            if (players[i].keyboardStates != null && players[i].keyboardStates.keyDown(e.keyCode)) {
                break;
            }
        }
    }

    onKeyUp(e: KeyboardEvent) {
        if (!this.gameOn()) {
            return;
        }

        var players = this.currentLocalPlayers();
        for (var i = 0; i < players.length; i++) {
            if (players[i].keyboardStates != null && players[i].keyboardStates.keyUp(e.keyCode)) {
                break;
            }
        }
    }

	private searchForGame() {
		this.myHub = $.connection.myHub;

		this.myHub.client.initGame = (message) => {
			this.handleInitGame(message);
		};
        this.myHub.client.newGameStart = (message) => {
			this.handleNewGameStart(message);
		};
		this.myHub.client.endGame = (message) => {
			utils.appendNewLine(this.textArea, message);
			this.currentPlayers.removeAll();
			window.cancelAnimationFrame(this.stopMain);
			this.gameOn(false);
			this.myHub.server.searchForGame();
            this.appendLine('Searching for game...');
		};
		this.myHub.client.updateGame = (message) => {
			this.handleServerUpdate(message);
		};

		$.connection.hub.error(error => {
			this.appendLine('SignalR error: ' + error)
		});

		$.connection.hub.disconnected(() => {
			this.appendLine('SignalR disconnected');
		});

		$.connection.hub.reconnecting(() => {
			this.appendLine('SignalR reconnecting');
		});

		$.connection.hub.start().done(() => {
			this.gameOn(false);
			this.connectionID = $.connection.hub.id;
            this.myHub.server.searchForGame();
            this.appendLine('Searching for game...');
		});
	}

	private handleInitGame(initGameEntity: SPATest.ServerCode.InitGameEntity) {
		this.appendLine('InitGame recived');
		this.gameOn(false);
        this.currentPlayers.removeAll();
        this.networkHandler = new NetworkHandler();

		initGameEntity.players.forEach(player => {
			var clientPlayer: ClientPlayer;
			if (player.connectionId === this.connectionID) {				
				clientPlayer = new ClientPlayer(player, KeyboardGroup.WSAD, true);
				this.appendLine('YOU ARE: <b><span style="color:' + clientPlayer.color + '">' + clientPlayer.color.toUpperCase() + '</span></b>');
				this.currentLocalPlayers([clientPlayer]);
			} else {
				clientPlayer = new ClientPlayer(player, null, false);
			}
			this.currentPlayers.push(clientPlayer);
        });
        this.initMap(initGameEntity.map);
		this.main(performance.now());
		this.myHub.server.sendReady();
		this.appendLine('SendReady sent');
	}

	private handleNewGameStart(newGameStartEntity: SPATest.ServerCode.NewGameStartEntity) {
		var gameStartInSeconds = moment.duration(moment(utils.getFormatedDateWithTime(newGameStartEntity.startTime)).diff(moment())).asSeconds();
		this.appendLine('Game start in ' + gameStartInSeconds + ' seconds');

		setTimeout(() => {
			this.appendLine('Game start!');
			this.gameOn(true);
		}, gameStartInSeconds * 1000);
	}

	private handleServerUpdate(updateGameEntity: SPATest.ServerCode.UpdateGameEntity) {
		if (updateGameEntity == null) {
			this.appendLine('<b><span style="color: red">ERROR - UpdateGameEntity was null!</span></b>');
			return;
		}

		updateGameEntity.players.forEach(player => {
			var playerWithId = this.currentPlayers().filter(currPlayer => {
				return currPlayer.connectionId === player.connectionId;
            });

            var isLocalPlayer = this.currentLocalPlayers().some(localPlayer => {
                return localPlayer.connectionId === player.connectionId;
            });

            if (playerWithId.length > 0 && ((!isLocalPlayer && playerWithId[0].latestFrameUpdate < player.latestFrameUpdate) || !this.networkHandler.checkUpdateFromServer(player))) {
                playerWithId[0].position = player.position;
            }
            this.currentMap().addMapPart(player);

            if (!player.isAlive && player.isAlive !== playerWithId[0].isAlive) {
                playerWithId[0].isAlive = false;
                this.appendLine('Player: ' + player.color + ' died!');
                this.gameOn(false);
            }
		});
	}

    private main = (tFrame: number) => {
        this.stopMain = window.requestAnimationFrame(this.main);
        var nextTick = this.lastTick + this.tickLength;
        var numTicks = 0;

        if (tFrame > nextTick) {
            var timeSinceTick = tFrame - this.lastTick;
            numTicks = Math.floor(timeSinceTick / this.tickLength);
        }

        this.queueUpdates(numTicks);
        this.redrawCanvas(tFrame);
        this.lastRender = tFrame;

        this.currentLocalPlayers().forEach(localPlayer => {
            var updateObj = < SPATest.ServerCode.SendUpdateGameEntity > {
                player: localPlayer,
                frame: tFrame
            };

            this.myHub.server.sendUpdate(updateObj).done(() => {
                this.networkHandler.addUpdate(updateObj);
            }).fail(error => {
                this.appendLine('<b> <span style="color: red"> ERROR - SendUpdateGameEntity failed!</span></b> - ' + error);
            });
        });
    }

    private queueUpdates(numTicks: number) {
        for (var i = 0; i < numTicks; i++) {
            this.lastTick = this.lastTick + this.tickLength;
            this.update(Math.floor(this.lastTick));
        }
    }

    private update(lastTick: number) {
        if (this.gameOn() === false) {
            return;
        }

        var map = this.currentMap();
        this.currentPlayers().forEach(player => {
            player.update(this.ctx, map, this.tickLength);
		});
    }

    private redrawCanvas(tFrame: number) {
        var deltaTick = tFrame - this.lastTick;
        Rendering.render(this.ctx, this.currentPlayers(), deltaTick, this.currentMap());
    }

    private initCanvas() {
        this.ctx.canvas.tabIndex = 1000;
        this.ctx.canvas.style.outline = "none";

		window.addEventListener("blur",(e) => { this.onBlur(); }, false);
        window.addEventListener("keydown", (e) => { this.onKeyDown(e); }, false);
        window.addEventListener("keyup", (e) => { this.onKeyUp(e); }, false);

        this.appendLine('Canvas init done');
    }

    private initMap(map: SPATest.ServerCode.Map) {
        var size = <SPATest.ServerCode.Size>{
            height: this.ctx.canvas.height,
            width: this.ctx.canvas.width
        };
        this.currentMap(new ClientMap(map));
    }

    private initLocalPlayers() {
        //Player 1
		var serverPlayer1 = <SPATest.ServerCode.Player>{
			team: SPATest.ServerCode.Team.BLUE,
			position: this.currentMap().getRandomStartPosition()
		};
        var player1 = new ClientPlayer(serverPlayer1, KeyboardGroup.WSAD, true);

        //Player 2
		var serverPlayer2 = <SPATest.ServerCode.Player>{
			team: SPATest.ServerCode.Team.RED,
			position: this.currentMap().getRandomStartPosition()
		};
        var player2 = new ClientPlayer(serverPlayer2, KeyboardGroup.Arrows, true);
        var settings = <SPATest.ServerCode.Map> {
            mapSize: <SPATest.ServerCode.Size>{
                height: 500,
                width: 1000
            },
            playerSize: 10,
            startPositionPadding: 10,
            tick: 0
        };

        this.initMap(settings);
        this.currentPlayers([player1, player2]);
        this.currentLocalPlayers(this.currentPlayers());
    }

	private appendLine(message: string) {
		utils.appendNewLine(this.textArea, message);
	}
}
export = Game;

File: AchtungLunneSpa/ServerCode/Game.cs
using System;
using SPATest.Hubs;
using System.Timers;
using System.Threading;

namespace SPATest.ServerCode
{
	public class Game
	{
		public string GroupReference { get; set; }
		public dynamic GroupManager { get; }
		public Map CurrentMap { get; }
		public Player player1 { get; }
		public Player player2 { get; }

		private System.Timers.Timer timer;
        private object updateLock = new object();

		public Game(Player player1, Player player2, MyHub myHub)
		{
			GroupReference = player1.ConnectionId + "-" + player2.ConnectionId + DateTime.Now.ToString();
			CurrentMap = new Map();

			this.player1 = player1;
			this.player1.GameGroupID = GroupReference;
			this.player1.Team = Team.RED;
            this.player1.Color = "red";
			

			this.player2 = player2;
			this.player2.Team = Team.BLUE;
            this.player2.Color = "blue";
            this.player2.GameGroupID = GroupReference;
			

			myHub.Groups.Add(player1.ConnectionId, GroupReference);
			myHub.Groups.Add(player2.ConnectionId, GroupReference);
			GroupManager = myHub.Clients.Group(GroupReference);
		}

		public void InitGame()
		{
            CurrentMap.ResetMapParts();
            player1.Position = CurrentMap.GetRandomPosition();
            player1.LastPositionVector = new BoundingBox(player1.Position, new Vector2D() { X = player1.Position.X + CurrentMap.PlayerSize, Y = player1.Position.Y + CurrentMap.PlayerSize });
            player1.IsAlive = true;

            player2.Position = CurrentMap.GetRandomPosition();
            player2.LastPositionVector = new BoundingBox(player2.Position, new Vector2D() { X = player2.Position.X + CurrentMap.PlayerSize, Y = player2.Position.Y + CurrentMap.PlayerSize });
            player2.IsAlive = true;
            GroupManager.initGame(new InitGameEntity { Map = CurrentMap, Players = new Player[] { player1, player2 } });
		}

		public void ReadyRecived(Player player)
		{
			player.IsReady = true;
			if (player1.IsReady && player2.IsReady)
			{
				StartGame();
			}
		}

		public void StartGame()
		{
			if (timer != null)
			{
				timer.Dispose();
			}

			timer = new System.Timers.Timer(33);
			timer.Elapsed += new ElapsedEventHandler(SendUpdate);
			GroupManager.newGameStart(new NewGameStartEntity() { StartTime = DateTime.Now.AddSeconds(5) });
			timer.Enabled = true;
		}

		public void EndGame(MyHub myHub)
		{
			if (timer != null)
			{
				timer.Dispose();
			}

			GroupManager.endGame("end - " + GroupReference);
			myHub.Groups.Remove(player1.ConnectionId, GroupReference);
			myHub.Groups.Remove(player2.ConnectionId, GroupReference);
		}

        public void SendUpdate(object sender, ElapsedEventArgs e)
        {
            lock (updateLock)
            {
				UpdateMap(player1);
				UpdateMap(player2);
				CurrentMap.Tick++;

                var updateEntity = new UpdateGameEntity { Players = new Player[] { player1, player2 } };
                GroupManager.updateGame(updateEntity);
            }
        }

        public void UpdateRecived(string connectionID, SendUpdateGameEntity entity)
        {
            var player = GetPlayer(connectionID);
            if (player != null && entity.Frame > player.LatestFrameUpdate && player.Position.X != entity.Player.Position.X && player.Position.Y != entity.Player.Position.Y)
            {
                lock (updateLock)
                {
                    player.Position = entity.Player.Position;
                    player.LatestFrameUpdate = entity.Frame;
                    if(timer != null && timer.Enabled && !CurrentMap.AddMapPart(player.Position, player))
                    {
                        player.IsAlive = false;
                        timer.Enabled = false;
                        SendUpdate(this, EventArgs.Empty as ElapsedEventArgs);
                    }
                }
            }
        }

        public Player GetPlayer(string connectionID)
		{
			if (player1.ConnectionId == connectionID)
			{
				return player1;
			}
			else if (player2.ConnectionId == connectionID)
			{
				return player2;
			}
			else
			{
				return null;
			}
		}

		public void UpdateMap(Player player)
		{
			
		}
	}
}